File: backend/internal/models/train_status.go
package models

import "time"

type TrainStatus struct {
	TrainNumber  string    `json:"train_number" db:"train_number"`
	CurrentLat   float64   `json:"current_lat" db:"current_lat"`
	CurrentLon   float64   `json:"current_lon" db:"current_lon"`
	DelayMinutes int       `json:"delay_minutes" db:"delay_minutes"`
	LastFetchedAt time.Time `json:"last_fetched_at" db:"last_fetched_at"`
}

type Breadcrumb struct {
	ID         int64     `json:"id" db:"id"`
	YatraID    int64     `json:"yatra_id" db:"yatra_id"`
	Lat        float64   `json:"lat" db:"lat"`
	Lon        float64   `json:"lon" db:"lon"`
	SnappedLat float64   `json:"snapped_lat" db:"snapped_lat"`
	SnappedLon float64   `json:"snapped_lon" db:"snapped_lon"`
	Timestamp  time.Time `json:"timestamp" db:"timestamp"`
}

type LocationUpdate struct {
	Lat float64 `json:"lat" binding:"required"`
	Lon float64 `json:"lon" binding:"required"`
}

File: app/lib/models/train_status.dart
class TrainStatus {
  final String trainNumber;
  final double currentLat;
  final double currentLon;
  final int delayMinutes;
  final DateTime lastFetchedAt;

  TrainStatus({
    required this.trainNumber,
    required this.currentLat,
    required this.currentLon,
    required this.delayMinutes,
    required this.lastFetchedAt,
  });

  factory TrainStatus.fromJson(Map<String, dynamic> json) {
    return TrainStatus(
      trainNumber: json['train_number'] as String,
      currentLat: (json['current_lat'] as num).toDouble(),
      currentLon: (json['current_lon'] as num).toDouble(),
      delayMinutes: json['delay_minutes'] as int,
      lastFetchedAt: DateTime.parse(json['last_fetched_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'train_number': trainNumber,
      'current_lat': currentLat,
      'current_lon': currentLon,
      'delay_minutes': delayMinutes,
      'last_fetched_at': lastFetchedAt.toIso8601String(),
    };
  }

  bool get isDelayed => delayMinutes > 0;

  String get delayText {
    if (delayMinutes <= 0) return 'On Time';
    if (delayMinutes < 60) return '$delayMinutes min late';
    final hours = delayMinutes ~/ 60;
    final mins = delayMinutes % 60;
    return '${hours}h ${mins}m late';
  }
}
